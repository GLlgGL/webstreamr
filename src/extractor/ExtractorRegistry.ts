import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Context, CountryCode, UrlResult } from '../types';
import { isExtractorDisabled } from '../utils';
import { Extractor } from './Extractor';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult[]>;

  public constructor(logger: winston.Logger, extractors: Extractor[]) {
    this.logger = logger;
    this.extractors = extractors;
    this.urlResultCache = new TTLCache({ max: 4096 });
  }

  public async handle(ctx: Context, url: URL, countryCode: CountryCode, title?: string | undefined): Promise<UrlResult[]> {
    const extractor = this.extractors.find(extractor => !isExtractorDisabled(ctx.config, extractor) && extractor.supports(ctx, url));
    if (!extractor) {
      return [];
    }

    const normalizedUrl = extractor.normalize(url);
    const cacheKey = extractor.viaMediaFlowProxy
      ? `${extractor.id}_${normalizedUrl}_${ctx.config.mediaFlowProxyUrl}`
      : `${extractor.id}_${normalizedUrl}`;

    let urlResults = this.urlResultCache.get(cacheKey) ?? [];
    if (this.urlResultCache.has(cacheKey)) {
      return urlResults.map(urlResult => ({ ...urlResult, ttl: this.urlResultCache.getRemainingTTL(cacheKey) }));
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    urlResults = await extractor.extract(ctx, normalizedUrl, countryCode, title);

    if (!urlResults.some(urlResult => urlResult.error) && extractor.ttl) {
      this.urlResultCache.set(cacheKey, urlResults, { ttl: extractor.ttl });
    }

    return urlResults;
  };
}
