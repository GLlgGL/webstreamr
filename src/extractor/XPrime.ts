import { Context, CountryCode, Format, UrlResult } from '../types';
import { Extractor } from './Extractor';

interface XPrimePrimeboxResponsePartial {
  streams: Record<string, string>;
  series_info?: {
    episode: number;
    season: number;
    title: string;
    year: string;
  };
  title?: string;
}

export class XPrime extends Extractor {
  public readonly id = 'xprime';

  public readonly label = 'XPrime';

  public supports(_ctx: Context, url: URL): boolean {
    return null !== url.host.match(/xprime/);
  }

  protected async extractInternal(ctx: Context, url: URL, countryCode: CountryCode): Promise<UrlResult[]> {
    const urlResults: UrlResult[] = [];

    const referer = url.protocol + '//' + url.hostname.split('.').slice(-2).join('.'); // Strip subdomains

    if (url.href.includes('primebox')) {
      const jsonResponse = JSON.parse(await this.fetcher.text(ctx, url)) as XPrimePrimeboxResponsePartial;

      for (const [resolution, stream] of Object.entries(jsonResponse.streams)) {
        const url = new URL(stream);
        const title = jsonResponse.series_info
          ? `${jsonResponse.series_info.title} ${jsonResponse.series_info.season}x${jsonResponse.series_info.episode} (${jsonResponse.series_info.year})`
          : jsonResponse.title;

        urlResults.push({
          url,
          format: Format.mp4,
          label: `${this.label} Primebox`,
          sourceId: `${this.id}_${countryCode}_primebox`,
          ttl: this.ttl,
          meta: {
            countryCodes: [countryCode],
            bytes: parseInt((await this.fetcher.head(ctx, url, { headers: { Referer: referer }, minCacheTtl: this.ttl }))['content-length'] as string),
            height: parseInt(resolution),
            title,
          },
          requestHeaders: {
            Referer: referer,
          },
        });
      }
    }

    return urlResults;
  };
}
