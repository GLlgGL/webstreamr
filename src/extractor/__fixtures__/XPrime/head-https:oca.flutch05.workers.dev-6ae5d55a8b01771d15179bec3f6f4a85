{"accept-ranges":"bytes","access-control-allow-headers":"*","access-control-allow-origin":"https://xprime.tv","alt-svc":"h3=\":443\"; ma=86400","cache-control":"public, max-age=31536000, immutable","cdn-cache-control":"public, max-age=31536000, immutable","cf-ray":"97d1622e3ad44d26-FRA","connection":"close","content-length":"896638128","content-type":"video/mp4","date":"Wed, 10 Sep 2025 19:38:03 GMT","nel":"{\"success_fraction\":0,\"report_to\":\"cf-nel\",\"max_age\":604800}","report-to":"{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=FzrMqwYe6qwzeh76DC9F4xwUbJJIwR3NLN%2BXEdKyul1Lv6zBocsQrsZtj%2F3mRAtOkr1R3fTuSSVikTLF3GqLes6idwSj2KRFSNQYB2qg%2BaS72v2SjbfnuEWL0eroypJ2346tnFninrVROOjBDB4vkwOI2e0tOyQ%3D\"}],\"group\":\"cf-nel\",\"max_age\":604800}","server":"cloudflare","server-timing":"cfL4;desc=\"?proto=TCP&rtt=0&min_rtt=0&rtt_var=0&sent=0&recv=0&lost=0&retrans=0&sent_bytes=0&recv_bytes=0&delivery_rate=0&cwnd=0&unsent_bytes=0&cid=0ab2b8c6b809570f&ts=85&x=0\"","vary":"Accept-Encoding"}