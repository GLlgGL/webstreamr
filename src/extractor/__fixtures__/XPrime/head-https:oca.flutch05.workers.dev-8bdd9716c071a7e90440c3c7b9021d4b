{"accept-ranges":"bytes","access-control-allow-headers":"*","access-control-allow-origin":"https://xprime.tv","alt-svc":"h3=\":443\"; ma=86400","cache-control":"public, max-age=31536000, immutable","cdn-cache-control":"public, max-age=31536000, immutable","cf-ray":"97d1622daa5b4d26-FRA","connection":"close","content-length":"401394599","content-type":"video/mp4","date":"Wed, 10 Sep 2025 19:38:03 GMT","nel":"{\"success_fraction\":0,\"report_to\":\"cf-nel\",\"max_age\":604800}","report-to":"{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=gs1y2VcEagyahSUbca2EMr5ZaVaQapIixbj3Mcgu4FGR8iUd5Sc0z8rXXo%2B2gVs6VpYrYqpTwv9ErfpTGbPrUc1Sc%2FYWrqfvE49d3MXKzCXu8LiDLj6Ap1q732n8DNyM8CUnpTt5559%2Bs1OsGhHzsfUqo646vhk%3D\"}],\"group\":\"cf-nel\",\"max_age\":604800}","server":"cloudflare","server-timing":"cfL4;desc=\"?proto=TCP&rtt=0&min_rtt=0&rtt_var=0&sent=0&recv=0&lost=0&retrans=0&sent_bytes=0&recv_bytes=0&delivery_rate=0&cwnd=0&unsent_bytes=0&cid=ee1864c7dc839181&ts=77&x=0\"","vary":"Accept-Encoding"}