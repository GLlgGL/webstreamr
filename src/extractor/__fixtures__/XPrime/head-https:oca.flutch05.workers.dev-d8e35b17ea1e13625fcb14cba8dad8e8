{"accept-ranges":"bytes","access-control-allow-headers":"*","access-control-allow-origin":"https://xprime.tv","alt-svc":"h3=\":443\"; ma=86400","cache-control":"public, max-age=31536000, immutable","cdn-cache-control":"public, max-age=31536000, immutable","cf-ray":"97d1622f5bbf4d26-FRA","connection":"close","content-length":"871281904","content-type":"video/mp4","date":"Wed, 10 Sep 2025 19:38:04 GMT","nel":"{\"success_fraction\":0,\"report_to\":\"cf-nel\",\"max_age\":604800}","report-to":"{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=KlG4q7%2FQZ7c2zxVghnbfh2TtHUadFAQfG%2F2MRX9YZS6lceSdjj93MyYssn7%2FJcnAWVtcR%2BlQ%2FuQHz7lp0PW0%2B1WIfBXn5mbLxMotY7mWWBS58JCSTZlDc2huJ8MZUuFCQiLLRO2o3IJhPCOVK%2FtkTrVqthc57AY%3D\"}],\"group\":\"cf-nel\",\"max_age\":604800}","server":"cloudflare","server-timing":"cfL4;desc=\"?proto=TCP&rtt=0&min_rtt=0&rtt_var=0&sent=0&recv=0&lost=0&retrans=0&sent_bytes=0&recv_bytes=0&delivery_rate=0&cwnd=0&unsent_bytes=0&cid=32fc89755f2b0f98&ts=110&x=0\"","vary":"Accept-Encoding"}