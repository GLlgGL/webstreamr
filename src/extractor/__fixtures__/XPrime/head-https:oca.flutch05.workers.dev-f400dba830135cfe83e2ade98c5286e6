{"accept-ranges":"bytes","access-control-allow-headers":"*","access-control-allow-origin":"https://xprime.tv","alt-svc":"h3=\":443\"; ma=86400","cache-control":"public, max-age=31536000, immutable","cdn-cache-control":"public, max-age=31536000, immutable","cf-ray":"97d162302c654d26-FRA","connection":"close","content-length":"420170402","content-type":"video/mp4","date":"Wed, 10 Sep 2025 19:38:04 GMT","nel":"{\"success_fraction\":0,\"report_to\":\"cf-nel\",\"max_age\":604800}","report-to":"{\"endpoints\":[{\"url\":\"https:\\/\\/a.nel.cloudflare.com\\/report\\/v4?s=fmqjV4Q2%2FjwPS%2F78MMmXHqgH9ibjI9QnwLkXiaGHmstwNDlLe0qOxfoKb2DvobWxarVPzYAxoyB3GdKKZ3GO%2B4KnNX5MHkDVgVFLVhUB04lqEHJTHwb4aG34yMdI5uziLuLxjz2t7BIprl4UFt%2BOinEinax4WH8%3D\"}],\"group\":\"cf-nel\",\"max_age\":604800}","server":"cloudflare","server-timing":"cfL4;desc=\"?proto=TCP&rtt=0&min_rtt=0&rtt_var=0&sent=0&recv=0&lost=0&retrans=0&sent_bytes=0&recv_bytes=0&delivery_rate=0&cwnd=0&unsent_bytes=0&cid=935db57878fc96c4&ts=62&x=0\"","vary":"Accept-Encoding"}