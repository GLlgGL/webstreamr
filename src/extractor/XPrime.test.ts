import winston from 'winston';
import { createTestContext } from '../test';
import { FetcherMock } from '../utils';
import { ExtractorRegistry } from './ExtractorRegistry';
import { XPrime } from './XPrime';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const extractorRegistry = new ExtractorRegistry(logger, [new XPrime(new FetcherMock(`${__dirname}/__fixtures__/XPrime`))]);

const ctx = createTestContext();

describe('XPrime', () => {
  test('unknown backend is ignored', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://backend.xprime.tv/unknown?name=Superman&year=2025'))).toMatchSnapshot();
  });

  test('primebox Superman', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://backend.xprime.tv/primebox?name=Superman&year=2025'))).toMatchSnapshot();
  });

  test('primebox Alien: Earth', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://backend.xprime.tv/primebox?name=Alien%3A%20Earth&year=2025&season=1&episode=1'))).toMatchSnapshot();
  });

  test('primebox not found', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://backend.xprime.tv/primebox?name=Jurassic%20World%20Rebirth&year=2025'))).toMatchSnapshot();
  });
});
