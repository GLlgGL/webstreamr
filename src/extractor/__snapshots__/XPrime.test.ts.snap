// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`XPrime primebox Alien: Earth 1`] = `
[
  {
    "format": "mp4",
    "label": "XPrime Primebox",
    "meta": {
      "bytes": 871281904,
      "countryCodes": [
        "en",
      ],
      "height": 1080,
      "title": "Alien: Earth 1x1 (2025)",
    },
    "requestHeaders": {
      "Referer": "https://xprime.tv",
    },
    "sourceId": "xprime_en_primebox",
    "ttl": 21600000,
    "url": "https://oca.flutch09.workers.dev/?v=cenBDs9Hbnv6PuXYzNUtVR2ZZ2YtmVtQIt6JqHB1dsbz2Y%2BXC9UwPurnE%2BQNk27Yr1WeYeLOOazAnSL8cU0DvVfg4pwKsjHKWIJICiVlvv9ndXII0bR7eaIi%2B11NbAj72pFFryHZjO6MkMKEJJnF4G%2BH1BXIO1ALds7J1A7gj2vc7DbUWL8kPxwH1rMw7NAys20M5s3PsbhgSrMpguZSBm8X&headers=%7B%22origin%22%3A%22https%3A%2F%2Fmoviebox.ng%22%2C%22referer%22%3A%22https%3A%2F%2Fmoviebox.ng%22%7D&safe",
  },
  {
    "format": "mp4",
    "label": "XPrime Primebox",
    "meta": {
      "bytes": 420170402,
      "countryCodes": [
        "en",
      ],
      "height": 480,
      "title": "Alien: Earth 1x1 (2025)",
    },
    "requestHeaders": {
      "Referer": "https://xprime.tv",
    },
    "sourceId": "xprime_en_primebox",
    "ttl": 21600000,
    "url": "https://oca.flutch09.workers.dev/?v=oS2n5QbXtpa%2BWXMpIIi0SqeqadEyYEAg6AmHvpTa3jzooh6zxINciUJrR%2B6b1xLZtxbnbAM3mrGGXOa5yckYkNSC%2BKGEFSdofXw298vq3ssHpI9LBN6XLvhtxnDfmub0QoMG%2Fkvgjhb8T%2FwiCHo9NeSX%2FetW6ANppW9jhZn0Hv3hsWdmES1jqj6nfIjQ1ur53htjr2vF7mZbZSsJ&headers=%7B%22origin%22%3A%22https%3A%2F%2Fmoviebox.ng%22%2C%22referer%22%3A%22https%3A%2F%2Fmoviebox.ng%22%7D&safe",
  },
]
`;

exports[`XPrime primebox Superman 1`] = `
[
  {
    "format": "mp4",
    "label": "XPrime Primebox",
    "meta": {
      "bytes": 401394599,
      "countryCodes": [
        "en",
      ],
      "height": 480,
      "title": "Superman",
    },
    "requestHeaders": {
      "Referer": "https://xprime.tv",
    },
    "sourceId": "xprime_en_primebox",
    "ttl": 21600000,
    "url": "https://oca.flutch09.workers.dev/?v=3RVZDNxcKNs6%2BqcdL1bDCgq9Yp1iPYqb6a9yToLq66vlzu08R1Hz7jiKx4OBH6YzJBcK7skuiDGDwIpro78ReVjMLtMHCfa0fBo0aJ9RYULOUMkLrZxjzSTXvo5Jss1Jiss8e9Z8pFdlC1VqwAUm6G9mx0u41dFLJ0GTJrwCesPe4zflbL8I76WkWYKBMT2MwHvDTIO4r3V0nf8BsRswsNJZ&headers=%7B%22origin%22%3A%22https%3A%2F%2Fmoviebox.ng%22%2C%22referer%22%3A%22https%3A%2F%2Fmoviebox.ng%22%7D&safe",
  },
  {
    "format": "mp4",
    "label": "XPrime Primebox",
    "meta": {
      "bytes": 896638128,
      "countryCodes": [
        "en",
      ],
      "height": 720,
      "title": "Superman",
    },
    "requestHeaders": {
      "Referer": "https://xprime.tv",
    },
    "sourceId": "xprime_en_primebox",
    "ttl": 21600000,
    "url": "https://oca.flutch09.workers.dev/?v=gUbhHOwpFIqajH6NnoH%2BYE9AWORUUogamkdYQedBfKi5eQtyniWtxmI1%2B5Swyl1ThC6vM0qHl53s0Q7j%2BDaWFhVRJQL%2BJRxbuOL1iCz1w5xVPLDU%2FTJ3zhf9EbULPGbbee14kdNYNc5%2FlF2sNe6RiUupsvhlj%2BjcgsT8uypX%2BudVyw1jVEUVogufPnc0lSrfoE98IZlOgdjBZKE3%2Fcw3erD2&headers=%7B%22origin%22%3A%22https%3A%2F%2Fmoviebox.ng%22%2C%22referer%22%3A%22https%3A%2F%2Fmoviebox.ng%22%7D&safe",
  },
]
`;

exports[`XPrime primebox not found 1`] = `[]`;

exports[`XPrime unknown backend is ignored 1`] = `[]`;
